File: structural/dart/facade_structured/subsystems.dart
class CPU {
  void freeze() { print('CPU: Freezing...'); }
  void jump(int position) { print('CPU: Jumping to position \$position'); }
  void execute() { print('CPU: Executing instructions...'); }
}

class Memory {
  void load(int position, String data) { print('Memory: Loading data \\'\$data\\' at position \$position'); }
}

class HardDrive {
  String read(int lba, int size) {
    print('HardDrive: Reading \$size bytes from LBA \$lba');
    return 'Boot Data';
  }
}

class Amplifier {
  void on() => print('Amplifier: ON');
  void off() => print('Amplifier: OFF');
  void setVolume(int level) => print('Amplifier: Volume set to \$level');
}

class Tuner {
  void on() => print('Tuner: ON');
  void off() => print('Tuner: OFF');
  void setFrequency(double freq) => print('Tuner: Frequency set to \$freq FM');
}

class DVDPlayer {
  void on() => print('DVD Player: ON');
  void off() => print('DVD Player: OFF');
  void play(String movie) => print('DVD Player: Playing \\'\$movie\\'');
}

class Projector {
  void on() => print('Projector: ON');
  void off() => print('Projector: OFF');
  void setInput(String input) => print('Projector: Input set to \$input');
}

class TheaterLights {
  void on() => print('Theater Lights: ON');
  void off() => print('Theater Lights: OFF');
  void dim(int level) => print('Theater Lights: Dimmed to \$level%');
}

class ProjectionScreen {
  void down() => print('Screen: DOWN');
  void up() => print('Screen: UP');
}

class PopcornPopper {
  void on() => print('Popcorn Popper: ON');
  void off() => print('Popcorn Popper: OFF');
  void pop() => print('Popcorn Popper: Popping popcorn!');
}

class InventorySystem {
  bool checkInventory(String item, int quantity) {
    print('Inventory: Checking \$quantity units of \$item');
    return true;
  }
  void updateInventory(String item, int quantity) {
    print('Inventory: Updating \$quantity units of \$item');
  }
}

class PaymentSystem {
  bool processPayment(double amount) {
    print('Payment: Processing payment of \$\amount');
    return true;
  }
  void refundPayment(double amount) {
    print('Payment: Refunding \$\amount');
  }
}

class ShippingSystem {
  void scheduleShipping(String address) { print('Shipping: Scheduling delivery to \$address'); }
  void cancelShipping() { print('Shipping: Cancelling delivery'); }
}

class NotificationSystem {
  void sendConfirmation(String email) { print('Notification: Sending confirmation to \$email'); }
  void sendCancellation(String email) { print('Notification: Sending cancellation to \$email'); }
}

File: structural/typescript/facade_structured/subsystems.ts
export class CPU {
  freeze(): void { console.log('CPU: Freezing...'); }
  jump(position: number): void { console.log(`CPU: Jumping to position \${position}`); }
  execute(): void { console.log('CPU: Executing instructions...'); }
}

export class Memory {
  load(position: number, data: string): void { console.log(`Memory: Loading data '\${data}' at position \${position}`); }
}

export class HardDrive {
  read(lba: number, size: number): string {
    console.log(`HardDrive: Reading \${size} bytes from LBA \${lba}`);
    return 'Boot Data';
  }
}

export class Amplifier {
  on(): void { console.log('Amplifier: ON'); }
  off(): void { console.log('Amplifier: OFF'); }
  setVolume(level: number): void { console.log(`Amplifier: Volume set to \${level}`); }
}

export class Tuner {
  on(): void { console.log('Tuner: ON'); }
  off(): void { console.log('Tuner: OFF'); }
  setFrequency(freq: number): void { console.log(`Tuner: Frequency set to \${freq} FM`); }
}

export class DVDPlayer {
  on(): void { console.log('DVD Player: ON'); }
  off(): void { console.log('DVD Player: OFF'); }
  play(movie: string): void { console.log(`DVD Player: Playing '\${movie}'`); }
}

export class Projector {
  on(): void { console.log('Projector: ON'); }
  off(): void { console.log('Projector: OFF'); }
  setInput(input: string): void { console.log(`Projector: Input set to \${input}`); }
}

export class TheaterLights {
  on(): void { console.log('Theater Lights: ON'); }
  off(): void { console.log('Theater Lights: OFF'); }
  dim(level: number): void { console.log(`Theater Lights: Dimmed to \${level}%`); }
}

export class Screen {
  up(): void { console.log('Screen: UP'); }
  down(): void { console.log('Screen: DOWN'); }
}

export class PopcornPopper {
  on(): void { console.log('Popcorn Popper: ON'); }
  off(): void { console.log('Popcorn Popper: OFF'); }
  pop(): void { console.log('Popcorn Popper: Popping popcorn!'); }
}

export class InventorySystem {
  checkInventory(item: string, quantity: number): boolean {
    console.log(`Inventory: Checking \${quantity} units of \${item}`);
    return true;
  }
  updateInventory(item: string, quantity: number): void {
    console.log(`Inventory: Updating \${quantity} units of \${item}`);
  }
}

export class PaymentSystem {
  processPayment(amount: number): boolean {
    console.log(`Payment: Processing payment of $\${amount}`);
    return true;
  }
  refundPayment(amount: number): void {
    console.log(`Payment: Refunding $\${amount}`);
  }
}

export class ShippingSystem {
  scheduleShipping(address: string): void { console.log(`Shipping: Scheduling delivery to \${address}`); }
  cancelShipping(): void { console.log('Shipping: Cancelling delivery'); }
}

export class NotificationSystem {
  sendConfirmation(email: string): void { console.log(`Notification: Sending confirmation to \${email}`); }
  sendCancellation(email: string): void { console.log(`Notification: Sending cancellation to \${email}`); }
}
